flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -236,6 +239,16 @@ public SchedulerBase(
         this.exceptionHistory =
                 new BoundedFIFOQueue<>(
                         jobMasterConfiguration.getInteger(WebOptions.MAX_EXCEPTION_HISTORY_SIZE));
+
+        this.finishingTasks = createFinishingTasks();
+    }
+
+    private Map<JobVertexID, BitSet> createFinishingTasks() {
+        Map<JobVertexID, BitSet> map = new HashMap<>();
+        for (ExecutionJobVertex vertex : getExecutionGraph().getAllVertices().values()) {
+            map.put(vertex.getJobVertexId(), new BitSet());
```

Comment:
Should we exclude ExecutionVertex that has been marked finished?

Otherwise, support the job is restarted after some tasks finished. Then checkpoint will not be triggered even if all tasks have finished.

Reply:
Actually, we don't have to. I suppose you are talking about a situation that the job is stopped with some subtasks finished and then restarted. In such cases, all tasks would be reinitializing after the job is restarted and then transform to `RUNNING` and `FINISHED`, also the `EndOfData` would be sent.
To verify it , I add a test case `ImmediateCheckpointingAfterAllTasksFinishedITCase#testRestoreAfterSomeTasksFinished`.

Reply:
Suppose all subtasks of a job vertex have finished and then the job is restarted, will JM still allocate a slot to deploy this job vertex? If yes, then it seems like a waste of slot resource and we probably should fix it.

Can you help provide the link to the code or doc that verifies the above behavior?

Reply:
Based on my test case, yes JM will allocate slots for finished tasks when restoring but I didn't find out the doc that states this point. Maybe the following stack trace can illustrate that the scheduler would allocate slots for all vertices of the jobgraph, regardless of whether it has been completed before. 
I will continue to delve deeper into the code to find the evidence, but I hope it doesn't block your review.
```
allocateSlotsFor:106, SlotSharingExecutionSlotAllocator (org.apache.flink.runtime.scheduler)
allocateSlotsFor:127, DefaultExecutionDeployer (org.apache.flink.runtime.scheduler)
allocateSlotsAndDeploy:98, DefaultExecutionDeployer (org.apache.flink.runtime.scheduler)
allocateSlotsAndDeploy:483, DefaultScheduler (org.apache.flink.runtime.scheduler)
scheduleRegion:287, PipelinedRegionSchedulingStrategy (org.apache.flink.runtime.scheduler.strategy)
accept:-1, 1191666286 (org.apache.flink.runtime.scheduler.strategy.PipelinedRegionSchedulingStrategy$$Lambda$1307)
forEach:1259, ArrayList (java.util)
maybeScheduleRegions:233, PipelinedRegionSchedulingStrategy (org.apache.flink.runtime.scheduler.strategy)
startScheduling:187, PipelinedRegionSchedulingStrategy (org.apache.flink.runtime.scheduler.strategy)
startSchedulingInternal:243, DefaultScheduler (org.apache.flink.runtime.scheduler)
startScheduling:623, SchedulerBase (org.apache.flink.runtime.scheduler)
startScheduling:1069, JobMaster (org.apache.flink.runtime.jobmaster)
startJobExecution:986, JobMaster (org.apache.flink.runtime.jobmaster)
onStart:432, JobMaster (org.apache.flink.runtime.jobmaster)
```

Reply:
Yes, JM will allocate slots for those vertices during restarting, regardless of whether it has been completed before. Those vertices would be initialized and reach the end of data immediately, then changed to FINISHED after performing a checkpoint. The test `CheckpointAfterAllTasksFinishedITCase#testRestoreAfterSomeTasksFinished` can prove it.